pkg/util/push/push.go

```diff
@@ -28,31 +49,232 @@ func Handler(maxRecvMsgSize int, sourceIPs *middleware.SourceIPExtractor, push F
 				logger = log.WithSourceIPs(source, logger)
 			}
 		}
-		var req cortexpb.PreallocWriteRequest
-		err := util.ParseProtoReader(ctx, r.Body, int(r.ContentLength), maxRecvMsgSize, &req, util.RawSnappy)
+
+		// follow Prometheus https://github.com/prometheus/prometheus/blob/main/storage/remote/write_handler.go
+		contentType := r.Header.Get("Content-Type")
+		if contentType == "" {
+			contentType = appProtoContentType
+		}
+
+		msgType, err := parseProtoMsg(contentType)
 		if err != nil {
-			level.Error(logger).Log("err", err.Error())
-			http.Error(w, err.Error(), http.StatusBadRequest)
+			level.Error(logger).Log("Error decoding remote write request", "err", err)
+			http.Error(w, err.Error(), http.StatusUnsupportedMediaType)
 			return
 		}
 
-		req.SkipLabelNameValidation = false
-		if req.Source == 0 {
-			req.Source = cortexpb.API
+		if msgType != config.RemoteWriteProtoMsgV1 && msgType != config.RemoteWriteProtoMsgV2 {
+			level.Error(logger).Log("Not accepted msg type", "msgType", msgType, "err", err)
+			http.Error(w, err.Error(), http.StatusUnsupportedMediaType)
+			return
 		}
 
-		if _, err := push(ctx, &req.WriteRequest); err != nil {
-			resp, ok := httpgrpc.HTTPResponseFromError(err)
-			if !ok {
-				http.Error(w, err.Error(), http.StatusInternalServerError)
+		enc := r.Header.Get("Content-Encoding")
+		if enc == "" {
+		} else if enc != string(remote.SnappyBlockCompression) {
+			err := fmt.Errorf("%v encoding (compression) is not accepted by this server; only %v is acceptable", enc, remote.SnappyBlockCompression)
+			level.Error(logger).Log("Error decoding remote write request", "err", err)
+			http.Error(w, err.Error(), http.StatusUnsupportedMediaType)
+			return
+		}
+
+		switch msgType {
+		case config.RemoteWriteProtoMsgV1:
+			var req cortexpb.PreallocWriteRequest
+			err := util.ParseProtoReader(ctx, r.Body, int(r.ContentLength), maxRecvMsgSize, &req, util.RawSnappy)
+			if err != nil {
+				level.Error(logger).Log("err", err.Error())
+				http.Error(w, err.Error(), http.StatusBadRequest)
+				return
+			}
+
+			req.SkipLabelNameValidation = false
+			if req.Source == 0 {
+				req.Source = cortexpb.API
+			}
+
+			if _, err := push(ctx, &req.WriteRequest); err != nil {
+				resp, ok := httpgrpc.HTTPResponseFromError(err)
+				if !ok {
+					http.Error(w, err.Error(), http.StatusInternalServerError)
+					return
+				}
+				if resp.GetCode()/100 == 5 {
+					level.Error(logger).Log("msg", "push error", "err", err)
+				} else if resp.GetCode() != http.StatusAccepted && resp.GetCode() != http.StatusTooManyRequests {
+					level.Warn(logger).Log("msg", "push refused", "err", err)
+				}
+				http.Error(w, string(resp.Body), int(resp.Code))
+			}
+		case config.RemoteWriteProtoMsgV2:
+			var req writev2.Request
+			err := util.ParseProtoReader(ctx, r.Body, int(r.ContentLength), maxRecvMsgSize, &req, util.RawSnappy)
```

Comment:
@alanprot @danielblando
I wonder if we want to introduce a feature flag to control the behavior for RW v2 request. We can either ignore the request or convert to v1 and in the future maybe just accept as is.

Reply:
Let's add a feature flag for the purpose of rollout. If RW 2.0 conversion is enabled right away, then Ingesters need to be rolled out first because of the protocol change to return stats. If we want to rollout Ingester and Distributor the same time then things can go wrong without a feature flag.